### ISSUE TITLE
weak_delegate False Positives

### ISSUE THREAD

**Issue by racer1988** [realm/SwiftLint#2786]
### New Issue Checklist

- [x] Updated SwiftLint to the latest version (SwiftLint 0.33.0)
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)
similar to #2019 

### Describe the bug

```
private lazy var tableViewDelegate : MyTableViewDelegate = {
        let delegate = MyTableViewDelegate()
        delegate.actionDelegate = self
        return delegate
    }()
```

where

`@objc class MyTableViewDelegate: NSObject, UITableViewDelegate`

this is not a delegate that needs to be weak. it's a table view delegate.

`tableView.delegate = tableViewDelegate`

where the delegate is not retained by the tableView so it mustn't be `weak`

##### Complete output when running SwiftLint, including the stack trace and command used

```bash
$ swiftlint lint file
```

Linting 'file.swift' (1/1)

file.swift:29:18: warning: Weak Delegate Violation: Delegates should be weak to avoid reference cycles. (weak_delegate)

### Environment

* SwiftLint version (run `swiftlint version` to be sure)?

``` ./swiftlint version
0.33.0
```

* Installation method used (Homebrew, CocoaPods, building from source, etc)?

`CocoaPods`

* Paste your configuration file:

```
disabled_rules:
  - line_length
  - variable_name
  - force_cast
  - control_statement
  - file_length
  - force_try
  - empty_enum_arguments
  - cyclomatic_complexity
  - type_body_length
  - function_body_length
  - type_name
  - vertical_whitespace
  - unused_closure_parameter
  - comma
  - legacy_constructor
  - opening_brace
  - trailing_semicolon
  - trailing_newline
  - implicit_getter
  - empty_parentheses_with_trailing_closure
  - valid_ibinspectable
  - todo
  - large_tuple
  - block_based_kvo
  - redundant_string_enum_value
  - syntactic_sugar
  - statement_position
  - empty_parameters
  - function_parameter_count
  - nesting
  - closing_brace
  - closure_parameter_position
  - shorthand_operator
  - trailing_whitespace
  - colon
  - mark
  - unused_optional_binding
  - redundant_discardable_let
  - unneeded_break_in_switch
  - multiple_closures_with_trailing_closure
  - fallthrough
  - for_where
  - notification_center_detachment

excluded: # paths to ignore during linting. Takes precedence over `included`.
  - Pods
```

* Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations)?

`no`

  If so, paste their relative paths and respective contents.

* Which Xcode version are you using (check `xcode-select -p`)?

`Version 10.2.1 (10E1001)`

* Do you have a sample that shows the issue? Run `echo "[string here]" | swiftlint lint --no-cache --use-stdin --enable-all-rules`
  to quickly test if your example is really demonstrating the issue. If your example is more
  complex, you can use `swiftlint lint --path [file here] --no-cache --enable-all-rules`.

`see decribe the bug section`

**Comment by NrmeenTomoum** [comment]
I got same Error 
Weak Delegate Violation: Delegates should be weak to avoid reference cycles. (weak_delegate)
@racer1988 you reached for any solution?

**Comment by racer1988** [comment]
@NrmeenTomoum yes, this is a "bug" in SwiftLint and I wanted to report it so that the rule itself can be modified to address this.

why the rule is triggered? I even saw this happens even if the class or variable are not called delegate

**Comment by marcelofabri** [comment]
SwiftLint currently doesn't know if a delegate property is meant to be set externally (so it should be `weak`) or if it's there to act as a delegate for another object.

You can workaround this by naming the property something that doesn't end with `delegate`. You can also disable the rule locally.

That said, we probably could add an exception for `private` properties in the rule.

**Comment by racer1988** [comment]
@marcelofabri I agree, however this is triggered even if the variable is not called `delegate`!

**Comment by marcelofabri** [comment]
It triggers if the variable ends with `delegate`

**Comment by robertmryan** [comment]
> That said, we probably could add an exception for private properties in the rule.

That’s a promising idea (though, that of course presumes that you haven’t accepted an external reference in the `init` method), e.g.

```swift
class Foo {
    private let delegate: FooDelegate?

    init(delegate: FooDelegate) {
        self.delegate = delegate
    }
}
```

---

While we’re thinking about alternatives, another option is to disable the rule if it is initialized in its own declaration, e.g., here are real-world examples: 

```swift
private let mapViewDelegate = MapViewDelegate()
```

or

```swift
private let textFieldDelegate = PhoneNumberTextFieldDelegate()
```

It seems exceedingly confusing to produce warning says “should be weak” when (a) this is a constant and cannot be `weak`; and (b) if you did make this a `weak var`, this delegate object to immediately be immediately deallocated. It seems that if it is being initialized in its declaration, that’s a good candidate for not applying the `weak_delegate` rule.

**Comment by marcelofabri** [comment]
> While we’re thinking about alternatives, another option is to disable the rule if it is initialized in its own declaration

Or if it's a `let` declaration 🤔

**Comment by robertmryan** [comment]
Yeah, perhaps. When I first drafted my comment, I was considering that, but stepped back from it on further reflection. Consider:

```swift
class Foo {
    private let delegate: FooDelegate

    init(delegate: FooDelegate) {
        self.delegate = delegate
    }
}
```

In that case, `weak_delegate` rule might be appropriate, with perhaps the following being better:

```swift
class Foo {
    private weak var delegate: FooDelegate?

    init(delegate: FooDelegate) {
        self.delegate = delegate
    }
}
```

That’s why I gravitated to declaration followed by `=` and some initialization...

**Comment by mime29** [comment]
In our project, we have:
```
struct One {
    let delegate: OneTwoDelegate
    init(delegate: OneTwoDelegate) {
        self.delegate = delegate
    }
}
```
Can the rule be deactivated when it's happening in `struct` please?

**Comment by lawgimenez** [comment]
It's 2020. I'm still getting this false positive.

**Comment by rockshassa** [comment]
This is still an issue in 2021, can we get some sort of resolution here please?

**Comment by yonat** [comment]
Now that weak_delegate is autocorrected, this rule can causes errors. For example https://github.com/Alamofire/Alamofire/issues/3046 .

**Comment by hakkurishian** [comment]
2022 and still having this issue

**Comment by racer1988** [comment]
@marcelofabri I wanted to add, after 2 years 🤣 , that I tested your suggestion of not calling it delegate and I was still getting a false positive.
I think this checks both name and type

**Comment by jpsim** [comment]
I'm just going to remove this rule, it's been a constant source of confusion and frustration among SwiftLint users since its introduction.
